### PR TITLE
PE-2879

### PR DESCRIPTION
**Pull request by thegrego**
New features:
- `email` - taken from `UserDetails`
- `government_gateway:
    { 
        "user_id": // GGW credId 
        "gateway_token": // GGW Token 
        "roles": // GGW Credential Role 
        "affinity_group": // Individual, Organisation, Agent
    }` - taken from `UserDetails`

Majority of the code refactoring is done:
- Use of flatmap to bind some `Futures` and their dependencies.
- Send less request per endpoint like `auth` as lot of information needs data from `auth` so it makes sense to ask it once.
- Use build in JSON Formatter provide by Play.
- Unify `api` a bit.
- Clean the code a little.

### CODE REVIEW

File: app/uk/gov/hmrc/openidconnect/userinfo/services/UserInfoService.scala [hunk 1]
```diff
@@ -47,36 +45,41 @@ trait LiveUserInfoService extends UserInfoService {
...
     scopes flatMap { scopes =>
-      val scopesForNino = Set("profile", "address", "openid:gov-uk-identifiers")
-      if ((scopesForNino -- scopes).size != scopesForNino.size) {
-        authConnector.fetchNino() onComplete {
-          case Success(nino) => promiseNino success nino
-          case Failure(exception) => throw exception
-        }
-      } else promiseNino success None
+      def getMaybeForScopes[T](maybeScopes: Set[String], allScopes: Set[String], f: => Future[Option[T]]): Future[Option[T]] = {
+        if ((maybeScopes -- allScopes).size != maybeScopes.size) f
```
Comment (thegrego) [inline app/uk/gov/hmrc/openidconnect/userinfo/services/UserInfoService.scala]: Unfortunately `intersects` gives the intersection of the sets. What I need here is only the difference as I am checking if at least one of registered scopes is also in maybe scopes.
The following code shows the difference between `intersects` and `--` on sets:
```
val maybeScopes = Set("a", "c", "d")
val registeredScopes1 = Set("a", "b")
val registeredScopes2 = Set("a", "c", "b", "f")
val registeredScopes3 = Set("a", "c", "d")
val registeredScopes4 = Set("a", "b", "c", "d")
val registeredScopes5 = Set("e", "f", "g")
   
maybeScopes -- registeredScopes1 // gives Set(c, d)
maybeScopes -- registeredScopes2 // gives Set(d)
maybeScopes -- registeredScopes3 // gives Set()
maybeScopes -- registeredScopes4 // gives Set()
maybeScopes -- registeredScopes5 // gives Set(a, c, d)

maybeScopes intersect registeredScopes1 // gives Set(a)
maybeScopes intersect registeredScopes2 // gives Set(a, c)
maybeScopes intersect registeredScopes3 // gives Set(a, c, d)
maybeScopes intersect registeredScopes4 // gives Set(a, c, d)
maybeScopes intersect registeredScopes5 // gives Set()
```

File: app/uk/gov/hmrc/openidconnect/userinfo/connectors/DesConnector.scala [hunk 2]
```diff
@@ -36,11 +36,12 @@ trait DesConnector {
...
   val desBearerToken: String
 
-  def fetchUserInfo(nino: Option[Nino])(implicit hc: HeaderCarrier): Future[Option[DesUserInfo]] = {
+  def fetchUserInfo(authority: Authority)(implicit hc: HeaderCarrier): Future[Option[DesUserInfo]] = {
     val newHc = hc.copy(authorization = Some(Authorization(s"Bearer $desBearerToken"))).withExtraHeaders("Environment" -> desEnvironment)
 
-    nino map { ninoValue =>
-      val url = s"$serviceUrl/pay-as-you-earn/individuals/${withoutSuffix(ninoValue.nino)}"
+    authority.nino map { ninoString =>
+      val nino = Nino(ninoString) // validates and handles invalid nino value
```
Comment (thegrego) [inline app/uk/gov/hmrc/openidconnect/userinfo/connectors/DesConnector.scala]: Done.

File: app/uk/gov/hmrc/openidconnect/userinfo/domain/GovernmentGatewayDetails.scala [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.openidconnect.userinfo.domain
+
+import uk.gov.hmrc.play.http.Token
+
+case class GovernmentGatewayDetails(
+                        user_id: Option[String],
+                        gateway_token: Option[Token],
```
Comment (thegrego) [inline app/uk/gov/hmrc/openidconnect/userinfo/domain/GovernmentGatewayDetails.scala]: Ugly but those fields are not used in the code - only in tests but I think this is acceptable. They just represent the JSON so keeping them like this lets me to use built-in JSON formatter from play with one line of code. If I change the fields names to cameCase I will have to provide much more code to map them to different JSON names.

File: test/unit/uk/gov/hmrc/openidconnect/userinfo/connectors/AuthConnectorSpec.scala [hunk 4]
```diff
@@ -18,70 +18,75 @@ package unit.uk.gov.hmrc.openidconnect.userinfo.connectors
...
+//    "return None when authority's NINO is not in the response" in new TestAuthConnector(wiremockBaseUrl) {
+//      given().get(urlPathEqualTo("/auth/authority")).returns("""{"credentialStrength":"weak"}""")
+//      fetchNino().futureValue shouldBe None
+//    }
+//
+//    "return None when auth request fails" in new TestAuthConnector(wiremockBaseUrl) {
+//      given().get(urlPathEqualTo("/auth/authority")).returns(500)
+//      fetchNino().futureValue shouldBe None
+//    }
+//  }
```
Comment (thegrego) [inline test/unit/uk/gov/hmrc/openidconnect/userinfo/connectors/AuthConnectorSpec.scala]: Deleted.

File: app/uk/gov/hmrc/openidconnect/userinfo/services/UserInfoService.scala (hunk 1 above)
Comment (fenallen) [inline app/uk/gov/hmrc/openidconnect/userinfo/services/UserInfoService.scala]: This is exactly what you want isn't it? I don't see the problem.

Comment (thegrego) [inline app/uk/gov/hmrc/openidconnect/userinfo/services/UserInfoService.scala]: Dumb me! `intersect` is what I want here.

**Pull request merged**